### PR TITLE
Table Cell Selection

### PR DESCRIPTION
**Pull request by BryanValverdeU**
Add a new Plugin for Table Selection.

+Ability to vertical and horizontal selection inside of a table.
+Ability to copy and paste selection made with the plugin selection
+Two content edit features to handle DELETE and BACKSPACE Keys when table selection.

Pending changes ->

-  change API package to handle selection inside of table.
- Table Selector using DragAndDropHelper approach.

### CODE REVIEW

**Comment by JiuqingSong** [comment]
Not a must, but consider adding options and sample code for the new plugin in demo site:

image

File: packages/roosterjs-editor-plugins/lib/plugins/TableSelection/TableCellSelection.ts [hunk 1]
```diff
@@ -0,0 +1,850 @@
...
+        }
+        clearSelectedTableCells(event.clonedRoot);
+    }
+
+    //#region Key events
+    /**
+     * Handles the on key event.
+     * @param event the plugin event
+     */
+    private handleKeyDownEvent(event: PluginKeyDownEvent | PluginKeyUpEvent) {
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-plugins/lib/plugins/TableSelection/TableCellSelection.ts]: It seems some of the code from this function can be replaced with Region API.

const regions = editor.getSelectedRegions();
Then check if there are multiple regions, and if they are in the same table.

File: packages/roosterjs-editor-plugins/lib/plugins/TableSelection/TableCellSelection.ts [hunk 2]
```diff
@@ -0,0 +1,850 @@
...
+
+    private handleMouseUpEvent(event: PluginMouseUpEvent) {
+        this.range = this.editor.getSelectionRange();
+        if (
+            event.isClicking &&
+            this.range.collapsed &&
+            !this.tableSelection &&
+            event.rawEvent.which != RIGHT_CLICK
+        ) {
+            this.clearTableCellSelection();
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-plugins/lib/plugins/TableSelection/TableCellSelection.ts]: why?

File: packages/roosterjs-editor-plugins/lib/plugins/TableSelection/TableCellSelection.ts [hunk 3]
```diff
@@ -0,0 +1,850 @@
...
+        // If the event target is not contained in the editor.
+        if (
+            (this.lastTarget.lastChild == this.lastTarget.firstChild &&
+                getTagOfNode(this.lastTarget.lastChild) == 'TABLE' &&
+                getTagOfNode(this.lastTarget) == 'DIV') ||
+            !this.contentDiv.contains(this.lastTarget)
+        ) {
+            event.preventDefault();
+            return;
+        }
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-plugins/lib/plugins/TableSelection/TableCellSelection.ts]: First, we should only handle mousemove when left button is down, right?

File: packages/roosterjs-editor-plugins/lib/plugins/TableSelection/TableCellSelection.ts [hunk 4]
```diff
@@ -0,0 +1,850 @@
...
+        const isNewTDContainingFirstTable = safeInstanceOf(this.lastTarget, 'HTMLTableCellElement')
+            ? contains(this.lastTarget, this.firstTable)
+            : false;
+
+        if (this.firstTable && (targetIsValid || isNewTDContainingFirstTable)) {
+            //When starting selection inside of a table and ends inside of the same table.
+            this.selectionInsideTableMouseMove(event);
+        } else if (this.tableSelection) {
+            clearSelectedTableCells(this.contentDiv);
+            this.tableSelection = false;
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-plugins/lib/plugins/TableSelection/TableCellSelection.ts]: sees some of the code can be replaced with editor.getSelectedRegions()

File: packages/roosterjs-editor-dom/lib/table/VTable.ts [hunk 5]
```diff
@@ -471,6 +487,277 @@ export default class VTable implements Table {
...
+                callback(this.cells[indexY][indexX], indexX, indexY);
+            }
+        }
+    }
+
+    /**
+     * Remove the cells outside of the selection.
+     * @param outsideOfSelection whether to remove the cells outside or inside of the selection
+     */
+    removeCellsBySelection(outsideOfSelection: boolean = true) {
```
[the remark this answers is not in the document]
Comment (JiuqingSong) [reply packages/roosterjs-editor-dom/lib/table/VTable.ts:697]: Actually it seems all these functions can be moved to plugin folder. They are not likely to be called from other places, and actually we don't even want people call them from other places

File: packages/roosterjs-editor-core/lib/corePlugins/CopyPastePlugin.ts [hunk 6]
```diff
@@ -75,7 +76,16 @@ export default class CopyPastePlugin implements PluginWithState<CopyPastePluginS
...
+        const selection = this.editor.getSelectionRangeEx();
+        let originalRange: Range = new Range();
+
+        if (selection.type == SelectionRangeTypes.TableSelection && selection.vTable) {
+            originalRange.selectNode(selection.vTable.table);
+            this.editor.select(originalRange);
+        } else if (selection.type == SelectionRangeTypes.Normal) {
+            originalRange = selection.ranges[0];
+        }
+
```
[the remark this answers is not in the document]
Comment (BryanValverdeU) [reply packages/roosterjs-editor-core/lib/corePlugins/CopyPastePlugin.ts:88]: Just removed the code from 82 to 88 and it is working fine as you mentioned, thanks.

File: packages/roosterjs-editor-plugins/lib/plugins/TableSelection/TableCellSelection.ts (hunk 4 above)
Comment (BryanValverdeU) [reply packages/roosterjs-editor-plugins/lib/plugins/TableSelection/TableCellSelection.ts]: Added getSelectedRegions to the code as recommended.

File: packages/roosterjs-editor-plugins/lib/plugins/TableCellSelection/utils/highlightCellHandler.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+        (!element.dataset[tableCellSelectionCommon.TEMP_BACKGROUND_COLOR] ||
+            element.dataset[tableCellSelectionCommon.TEMP_BACKGROUND_COLOR] == '')
+    ) {
+        element.dataset[tableCellSelectionCommon.TEMP_BACKGROUND_COLOR] =
+            element.style.backgroundColor ?? element.style.background ?? '';
+    }
+    element.style.backgroundColor = tableCellSelectionCommon.HIGHLIGHT_COLOR;
+    element.classList.add(tableCellSelectionCommon.TABLE_CELL_SELECTED);
+
+    element.querySelectorAll('table').forEach(table => {
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-plugins/lib/plugins/TableCellSelection/utils/highlightCellHandler.ts:23]: Why need to call to all tables?

new VTable has some cost. So we should avoid use it too much

**Review by JiuqingSong — APPROVED** [review]

File: packages/roosterjs-editor-plugins/lib/plugins/TableCellSelection/utils/highlightCellHandler.ts (hunk 7 above)
Comment (BryanValverdeU) [reply packages/roosterjs-editor-plugins/lib/plugins/TableCellSelection/utils/highlightCellHandler.ts]: I agree VTable should not be required in this case.
Refactoring so we dont use VTable.

**Pull request merged**